Blog: keycloak (#122)
JankariTech/blog: src/assets/Keycloak/ocis-with-keycloak.md

---

> title: keycloak Server

Reviewer: if this blog is focused on keycloak with ocis, then the title should say something about ocis

authorName: Amrita Shrestha
authorAvatar: https://avatars.githubusercontent.com/u/54478846?s=100&v=4
authorLink: https://github.com/amrita-shrestha
createdAt: March 26, 2024
tags: owncloud infinte scale, ocis, ocis-with-keycloak
banner: https://blog.jankaritech.com/src/assets/keycloak/images/keycloak-with-ocis.png
---

> [In previous blog](src/assets/keycloak/keycloak.md), we explored the topic of Keycloak. In this current post, our focus shifts to discussing the setup of OCIS with Keycloak.

Reviewer: ```suggestion
[In the previous blog](src/assets/keycloak/keycloak.md), we explored the general functionality of Keycloak. In this current post, I will show how to setup of [OCIS](link to ocis) with Keycloak.
```

Reviewer: Do you expect people to know what ocis is?

Reviewer: Why would someone want to do that?

Reviewer 2: As a reader, I would be interested how I can setup Keycloak for my own project :)

## Keycloak with ocis

### Run services

> Before setting up `ocis with keycloak`, it's essential to deploy `keycloak and ocis`. This blog details the deployment of `ocis and keycloak` via `docker compose`.

Reviewer: keycloak and ocis are mentioned a lot, maybe we can rephrase a bit like
```suggestion
Before setting up oCIS with Keycloak, it's essential to deploy both services. In this blog I will explain how to do that using `docker compose`.
```

> Using `Traefik`, `ocis and keycloak` has been served using certificates.

Reviewer: how do the certificates serve ocis and keycloak?
Do you mean that you are using traefik to be able to access ocis and keycloak over secure TLS connections?

> You can either download the code from [GitHub](https://github.com/amrita-shrestha/keycloak-ocis/blob/master/compose.yaml) or utilize the code provided below:

Reviewer: what do I do after downloading the file?

      traefik.enable: true
      traefik.http.routers.ocis.tls: true
      traefik.http.routers.ocis.rule: PathPrefix(`/`)
      traefik.http.routers.ocis.entrypoints: ocis
      traefik.http.services.ocis.loadbalancer.server.port: 9200

  postgres:
    image: postgres:alpine
    volumes:
      - postgres_data:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: keycloak
      POSTGRES_USER: keycloak
      POSTGRES_PASSWORD: keycloak

  keycloak:
    image: quay.io/keycloak/keycloak:24.0.1
    command: ["start-dev", "--proxy=edge"]
    environment:
      KC_DB: postgres
      KC_DB_URL: "jdbc:postgresql://postgres:5432/keycloak"
      KC_DB_USERNAME: keycloak
      KC_DB_PASSWORD: keycloak
      KC_FEATURES: impersonation
      KEYCLOAK_ADMIN: admin
      KEYCLOAK_ADMIN_PASSWORD: admin
    extra_hosts:
      - host.docker.internal:host-gateway
    labels:
      traefik.enable: true
      traefik.http.routers.keycloak.tls: true
      traefik.http.routers.keycloak.rule: PathPrefix(`/`)
      traefik.http.routers.keycloak.entrypoints: keycloak
      traefik.http.services.keycloak.loadbalancer.server.port: 8080
    depends_on:
      - postgres

volumes:
  postgres_data:
```

> 

Reviewer: maybe some more explanation about what the docker file does, we want to empower people to do setups themself

> ### Manual configuration for ocis on keycloak

Reviewer: is there an other way than a "manual" configuration? If so describe it, if not, no need to say that this is "manual"

![client registration](src/assets/keycloak/images/client-registration.png)

- Client ID: `web`
![add client](src/assets/keycloak/images/client-registration-2.png)

- Root URL: `https://host.docker.internal:9200`
![add root url](src/assets/keycloak/images/client-registration-3.png)

5. Add realm roles:

- Role name: `ocisAdmin`
![add realm role](src/assets/keycloak/images/create-realm-role.png)

6. Create a new user:

- Username: `admin` and other info
![create user](src/assets/keycloak/images/create-user-1.png)
![add user information](src/assets/keycloak/images/create-user-2.png)

- Create password
![create password](src/assets/keycloak/images/create-user-3.png)
![setup password for user](src/assets/keycloak/images/create-user-4.png)

- Role Mapping: Assign `ocisAdmin` role
![assign role](src/assets/keycloak/images/assign-role-to-user.png)
![assign role to user](src/assets/keycloak/images/asign-role-to-user-2.png)

7. Update `roles` Client scope: `Client Scopes` -> `roles`
- `Mappers` -> `realm roles` -> `Token Claim Name=roles`
![update-roles](src/assets/keycloak/images/update-roles.png)

8. Go to url `https://host.docker.internal:9200/` reload if already opened

> ![login ocis with keycloak](src/assets/keycloak/images/ocis-keycloak-login.png)

Reviewer: I'm at the end of the blog post, but what did I achieve?

Reviewer 2: Short conclusion/summary would be nice. What is the end result after following these steps? What do I need to be aware of if I want to replicated them to my own project?
